src/verbs/queue_pair.rs

```diff
@@ -38,26 +42,240 @@ pub enum QueuePairState {
+        let mut qp_attr = ibv_qp_attr { ..attr.attr };
+        let ret = unsafe { ibv_modify_qp(self.qp().as_ptr(), &mut qp_attr as *mut _, attr.attr_mask.bits) };
+        if ret == 0 {
+            Ok(())
+        } else {
+            check_transition_mask(attr.attr_mask, self.state(), attr.attr.qp_state.into()).unwrap();
+            Err(format!("ibv_modify_qp failed, err={ret}"))
+        }
+    }
+
+    fn state(&self) -> QueuePairState {
+        unsafe { (*self.qp().as_ref()).state.into() }
+    }
+}
+
+// According to C standard, enums should be int, but Rust just uses whatever
+// type returned by Clang, which is uint on Linux platforms, so just cast it
+// into int.
+//
+// https://github.com/rust-lang/rust-bindgen/issues/1966
+//
+#[bitmask(i32)]
+#[bitmask_config(vec_debug)]
+pub enum QueuePairAttributeMask {
+    State = ibv_qp_attr_mask::IBV_QP_STATE.0 as _,
+    CurrentState = ibv_qp_attr_mask::IBV_QP_CUR_STATE.0 as _,
+    EnableSendQueueDrainedAyscNotify = ibv_qp_attr_mask::IBV_QP_EN_SQD_ASYNC_NOTIFY.0 as _,
+    AccessFlags = ibv_qp_attr_mask::IBV_QP_ACCESS_FLAGS.0 as _,
+    PartitionKeyIndex = ibv_qp_attr_mask::IBV_QP_PKEY_INDEX.0 as _,
+    Port = ibv_qp_attr_mask::IBV_QP_PORT.0 as _,
+    QueueKey = ibv_qp_attr_mask::IBV_QP_QKEY.0 as _,
+    AddressVector = ibv_qp_attr_mask::IBV_QP_AV.0 as _,
+    PathMtu = ibv_qp_attr_mask::IBV_QP_PATH_MTU.0 as _,
+    Timeout = ibv_qp_attr_mask::IBV_QP_TIMEOUT.0 as _,
+    RetryCount = ibv_qp_attr_mask::IBV_QP_RETRY_CNT.0 as _,
+    ResponderNotReadyRetryCount = ibv_qp_attr_mask::IBV_QP_RNR_RETRY.0 as _,
+    ReceiveQueuePacketSequenceNumber = ibv_qp_attr_mask::IBV_QP_RQ_PSN.0 as _,
+    MaxReadAtomic = ibv_qp_attr_mask::IBV_QP_MAX_QP_RD_ATOMIC.0 as _,
+    AlternativePath = ibv_qp_attr_mask::IBV_QP_ALT_PATH.0 as _,
+    MinResponderNotReadyTimer = ibv_qp_attr_mask::IBV_QP_MIN_RNR_TIMER.0 as _,
+    SendQueuePacketSequenceNumber = ibv_qp_attr_mask::IBV_QP_SQ_PSN.0 as _,
+    MaxDestinationReadAtomic = ibv_qp_attr_mask::IBV_QP_MAX_DEST_RD_ATOMIC.0 as _,
+    PathMigrationState = ibv_qp_attr_mask::IBV_QP_PATH_MIG_STATE.0 as _,
+    Capabilities = ibv_qp_attr_mask::IBV_QP_CAP.0 as _,
+    DestinationQueuePairNumber = ibv_qp_attr_mask::IBV_QP_DEST_QPN.0 as _,
+    RateLimit = ibv_qp_attr_mask::IBV_QP_RATE_LIMIT.0 as _,
+}
+
+// Define the required and optional mask according to the spec, so that we
+// could provide attr check for users. Furthermore, provide more useful
+// error messages.
+//
+// There is a corresponding table named qp_state_table in Linux kernel
+//
+// Ref: https://elixir.bootlin.com/linux/v6.10.9/source/drivers/infiniband/core/verbs.c#L1385
+//
+// We should consider using `std::mem::variant_count` here, after it stablized.
+//
+#[derive(Debug, Copy, Clone)]
+struct QueuePairStateTableEntry {
+    // whether this state transition is valid.
+    valid: bool,
+    req_mask: QueuePairAttributeMask,
```

Comment:
Prefer using full name to abbreviation; in this case `required_mask` and `optional_mask` are preferred.